src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1485,6 +1408,402 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
+            /* Deregister each alias (value) */
+            ucs_trace("%s: deregistered UMR mkey alias %p lkey 0x%x",
+                    uct_ib_device_name(ibdev), mkey_alias.cross_mr,
+                    mkey_alias.lkey);
+
+            status = uct_ib_mlx5_devx_obj_destroy(mkey_alias.cross_mr,
+                                                  "MKEY_ALIAS");
+            if (UCS_OK != status) {
+                ucs_error("%s: failed to deregister UMR alias %p lkey 0x%x: %m",
+                        uct_ib_device_name(ibdev), mkey_alias.cross_mr,
+                        mkey_alias.lkey);
+            }
+        });
+
+        kh_destroy(umr_mkey_map, md->umr_mkey_hash);
+        md->umr_mkey_hash = NULL;
+        ucs_trace("%s: destroyed umr_mkey_hash", uct_ib_device_name(ibdev));
+    }
+
+    if (!ucs_list_is_empty(&md->umr_mkey_pool)) {
+        ucs_info("%s: UMR mkey pool has %lu elements",
+                 uct_ib_device_name(ibdev),
+                 ucs_list_length(&md->umr_mkey_pool));
+    }
+
+    /* Destroy umr_mkey_pool */
+    ucs_list_for_each_safe(item, tmp, &md->umr_mkey_pool, super) {
+        /* Deregister each indirect mkey */
+        ucs_trace("%s: deregistered UMR mkey %p lkey 0x%x",
+                  uct_ib_device_name(ibdev), item->umr_mkey,
+                  item->umr_mkey->lkey);
+        if (mlx5dv_destroy_mkey(item->umr_mkey) != 0) {
+            ucs_error("%s: failed to destroy UMR mkey %p lkey 0x%x: %m",
+                      uct_ib_device_name(ibdev), item->umr_mkey,
+                      item->umr_mkey->lkey);
+        }
+
+        ucs_list_del(&item->super);
+        free(item);
+        ucs_trace("%s: deleted UMR mr from pool (%ld)",
+                  uct_ib_device_name(ibdev),
+                  ucs_list_length(&md->umr_mkey_pool));
+    }
+
+    ucs_trace("%s: destroyed UMR mkey pool", uct_ib_device_name(ibdev));
+
+    if (NULL != md->umr_qp) {
+        if (ibv_destroy_qp(md->umr_qp) != 0) {
+            ucs_error("%s: failed to destroy UMR QP: %m",
+                    uct_ib_device_name(ibdev));
+        }
+
+        md->umr_qp = NULL;
+    }
+
+    if (NULL != md->umr_cq) {
+        if (ibv_destroy_cq(md->umr_cq) != 0) {
+            ucs_error("%s: failed to destroy UMR CQ: %m",
+                    uct_ib_device_name(ibdev));
+        }
+
+        md->umr_cq = NULL;
+    }
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_umr_mkey_invalidate(uct_ib_mlx5_md_t *md,
+                                     uct_ib_mlx5_devx_umr_t *umr)
+{
+    uint32_t mkey         = umr->umr_mkey->lkey;
+    struct ibv_wc wc      = {};
+    struct ibv_send_wr wr = {
+        .wr_id           = 0,
+        .next            = NULL,
+        .num_sge         = 0,
+        .opcode          = IBV_WR_LOCAL_INV,
+        .send_flags      = IBV_SEND_INLINE | IBV_SEND_SIGNALED,
+        .invalidate_rkey = mkey
+    };
+    int cq_ret;
+    struct ibv_send_wr *bad_wr;
+
+    if (ibv_post_send(md->umr_qp, &wr, &bad_wr) != UCS_OK) {
```

Comment:
need to check for `int` value defined by IB verbs rather than comparing with UCS codes (as it is not a UCX API function)